File: pgpainless-core/src/main/java/org/pgpainless/decryption_verification/IntegrityProtectedInputStream.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.decryption_verification;

import java.io.IOException;
import java.io.InputStream;
import javax.annotation.Nonnull;

import org.bouncycastle.openpgp.PGPEncryptedData;
import org.bouncycastle.openpgp.PGPException;
import org.pgpainless.exception.ModificationDetectionException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class IntegrityProtectedInputStream extends InputStream {

    private static final Logger LOGGER = LoggerFactory.getLogger(IntegrityProtectedInputStream.class);

    private final InputStream inputStream;
    private final PGPEncryptedData encryptedData;
    private final ConsumerOptions options;
    private boolean closed = false;

    public IntegrityProtectedInputStream(InputStream inputStream, PGPEncryptedData encryptedData, ConsumerOptions options) {
        this.inputStream = inputStream;
        this.encryptedData = encryptedData;
        this.options = options;
    }

    @Override
    public int read() throws IOException {
        return inputStream.read();
    }

    @Override
    public int read(@Nonnull byte[] b, int offset, int length) throws IOException {
        return inputStream.read(b, offset, length);
    }

    @Override
    public void close() throws IOException {
        if (closed) {
            return;
        }
        closed = true;

        if (encryptedData.isIntegrityProtected() && !options.isIgnoreMDCErrors()) {
            try {
                if (!encryptedData.verify()) {
                    throw new ModificationDetectionException();
                }
                LOGGER.debug("Integrity Protection check passed");
            } catch (PGPException e) {
                throw new IOException("Failed to verify integrity protection", e);
            }
        }
    }
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/decryption_verification/IntegrityProtectedInputStream.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.decryption_verification

import java.io.IOException
import java.io.InputStream
import org.bouncycastle.openpgp.PGPEncryptedData
import org.bouncycastle.openpgp.PGPException
import org.pgpainless.exception.ModificationDetectionException

class IntegrityProtectedInputStream(
    private val inputStream: InputStream,
    private val encryptedData: PGPEncryptedData,
    private val options: ConsumerOptions
) : InputStream() {
    private var closed: Boolean = false

    override fun read() = inputStream.read()

    override fun read(b: ByteArray, off: Int, len: Int) = inputStream.read(b, off, len)

    override fun close() {
        if (closed) return

        closed = true
        if (encryptedData.isIntegrityProtected && !options.isIgnoreMDCErrors()) {
            try {
                if (!encryptedData.verify()) throw ModificationDetectionException()
            } catch (e: PGPException) {
                throw IOException("Data appears to not be integrity protected.", e)
            }
        }
    }
}
